### PR TITLE
fp16 compute

### PR DESCRIPTION
**Pull request by anaruse**
This PR is related to fp16 compute with cuDNN 5 or above.

Current chainer supports fp16 I/O, meaning tensor data is kept as fp16 precision and compute is done at fp32 precision. This branch allows the compute at fp16 precision. If you already have models for fp16 I/O, usage of fp16 compute is very simple, just using **chainer.use_fp16_compute()** as below.

    with chainer.use_fp16_compute():
        loss = model(x, t)

Please also see the file [example/imagenet/alex.py](https://github.com/anaruse/chainer/blob/fp16_compute/examples/imagenet/alex.py) for an example of the usage.

Note 1: Computation of convolution gets done at fp16 precision using **chainer.user_fp16_compute()**, but computation of others such as activation and normalization stays at fp32 precision.

Note 2: You should use GPUs with compute capability 6.0 like Tesla P100 to test this branch, because fp16 compute is not fully supported by GPUs with compute capability other than 6.0.

### CODE REVIEW

**Comment by unnonouno** [comment]
I have a simple question. I feel it's enough to apply fp16 computation when input variables are fp16. Users can control computation behavior by changing input dtype. Is it a bad idea?

**Comment by anaruse** [comment]
Thank you for your comment.

I think following two options should be given to users: 1) **Pseudo fp16** (fp16 I/O and fp32 compute) and 2) **True fp16** (fp16 I/O and fp16 compute).

This is mainly because there is trade-off between them. True fp16 is faster than pseudo fp16 (when GPUs like Tesla P100 with full fp16 capability are used), but true fp16 might be less accurate than pseudo fp16. Why less accurete? In case of convolution used during forward propagation, each output number is an accumulation of CxHxW numbers (C: number of input channels, H: kernel height, W: kernel width). This accumulation is done at fp32 precision with pseudo fp16, and at fp16 precision with true fp16. Considering floating point's rounding error, fp32 accumulation can generate more accurate result then fp16 accumulation, especially when CxHxW is not small.

Then, which is better, pseudo fp16 or true fp16? Probably, it is model dependent. Pseudo fp16 might be good for some models, and true fp16 might be good for others (and both might be bad for the others..). So, this is why I think both two options are necessary for now.

**Comment by unnonouno** [comment]
So users can use fp16 computation without casting fp16 arrays to fp32, and you think cast is overhead. Is it right?
Does this problem occur on fp32 data and fp64 computation?

**Comment by anaruse** [comment]
That's right. You don't need to explicitly cast (or convert) fp16 arrays to fp32 in order to use **pseudo fp16**. The data conversion from fp16 to fp32 (and also from fp32 to fp16 after computation) is internally done by cuDNN. So, there are some overhead for the data conversion, but it is marginal as far as I know.
I think that **pseudo fp32** (fp32 data and fp64 compute) and **pseudo fp16** are alike in that point, though cuDNN does not have **pseudo fp32** mode.

File: chainer/function.py [hunk 1]
```diff
@@ -78,6 +78,24 @@ def force_backprop_mode():
     _thread_local.default_backprop = default
 
 
+@contextlib.contextmanager
+def use_fp16_compute():
+    """Enable fp16 compute for convolution related functions."""
+    default = getattr(_thread_local, 'fp16_compute_mode', False)
+    _thread_local.fp16_compute_mode = True
+    yield
```
Comment (unnonouno) [inline chainer/function.py]: Please use try-catch.

```
try:
    yield
finally:
    _thread_local.fp16_compute_mode = default
```

File: chainer/functions/connection/convolution_2d.py [hunk 2]
```diff
@@ -38,6 +41,15 @@ def __init__(self, stride=1, pad=0, use_cudnn=True, cover_all=False,
         self.cover_all = cover_all
         self.deterministic = deterministic
 
+        self.fp16_compute_mode = getattr(function._thread_local,
+                                         'fp16_compute_mode', False)
+        if self.fp16_compute_mode:
+            # fp16 compute is available on GPUs with compute capability 6.0.
+            if not _cc == '60':
+                msg = "fp16 compute is not supported by the GPU " + \
```
Comment (unnonouno) [inline chainer/functions/connection/convolution_2d.py]: Use single quotes for string literals.

File: chainer/functions/connection/convolution_2d.py [hunk 3]
```diff
@@ -38,6 +41,15 @@ def __init__(self, stride=1, pad=0, use_cudnn=True, cover_all=False,
         self.cover_all = cover_all
         self.deterministic = deterministic
 
+        self.fp16_compute_mode = getattr(function._thread_local,
+                                         'fp16_compute_mode', False)
+        if self.fp16_compute_mode:
+            # fp16 compute is available on GPUs with compute capability 6.0.
+            if not _cc == '60':
```
Comment (unnonouno) [inline chainer/functions/connection/convolution_2d.py]: ```if int(_cc) < 60``` ?

File: chainer/functions/connection/convolution_2d.py [hunk 4]
```diff
@@ -201,17 +220,43 @@ def backward_gpu(self, inputs, grad_outputs):
...
+                    if self.sx != 1 or self.sy != 1:
+                        conv_desc_bd = self.conv_desc_fp32
+
+                    try:
+                        # try to find algo for fp16 compute
+                        algo = libcudnn.getConvolutionBackwardDataAlgorithm(
+                            handle, self.filter_desc.value, gy_desc.value,
+                            conv_desc_bd.value, x_desc.value, _bwd_data_pref,
+                            workspace_size)
+                    except RuntimeError:
```
Comment (unnonouno) [inline chainer/functions/connection/convolution_2d.py]: When this error happens?

File: chainer/functions/connection/convolution_2d.py [hunk 5]
```diff
@@ -201,17 +220,43 @@ def backward_gpu(self, inputs, grad_outputs):
...
+                        # print("No algorithm found for fp16 compute (kh:{}, kw:{}, n:{}, c:{}, h:{}, w:{}, out_c:{}, out_h:{}, out_w:{})".format(kh, kw, n, c, h, w, out_c, out_h, out_w))  # NOQA
```
Comment (unnonouno) [inline chainer/functions/connection/convolution_2d.py]: Do we need to show warning?

File: chainer/functions/connection/convolution_2d.py (hunk 3 above)
Comment (anaruse) [inline chainer/functions/connection/convolution_2d.py]: It cannot be decided if a GPU has full FP16 capability from simple large/small relation of the compute capability number. Please refer to [this link](http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#arithmetic-instructions) about what GPU/compute capability has full FP16 capability.

File: chainer/functions/connection/convolution_2d.py (hunk 2 above)
Comment (anaruse) [inline chainer/functions/connection/convolution_2d.py]: Modified to use single quotes

File: chainer/functions/connection/convolution_2d.py (hunk 4 above)
Comment (anaruse) [inline chainer/functions/connection/convolution_2d.py]: It depends on cuDNN version when this happens. This issue was personally observed when cuDNN 5.1 was used with stride > 1.

File: chainer/functions/connection/convolution_2d.py (hunk 5 above)
Comment (anaruse) [inline chainer/functions/connection/convolution_2d.py]: I feel it would be better to show the warning, but most users may not like to see the warning like this. So, I will leave to you as to whether or not to show the warning.

File: chainer/function.py (hunk 1 above)
Comment (anaruse) [inline chainer/function.py]: modified to use try-catch.

File: chainer/functions/connection/convolution_2d.py (hunk 5 above)
Comment (unnonouno) [inline chainer/functions/connection/convolution_2d.py]: Chainer has debug mode. How about showing warning in the debug mode?

Comment (anaruse) [inline chainer/functions/connection/convolution_2d.py]: It's a good idea. changed to show the warning when debug mode is turned on.

**Comment by anaruse** [comment]
"rebase master" has been done to resolve come conflicts. I have confirmed that it can be build and run normally with cuDNN 5.1 and 6.0.
